### PR TITLE
Feature: add support for persisting folder expanded collapsed state

### PR DESCRIPTION
**Pull request by billypchan**
## What's new in this PR?

Support state persistency for folder collapsed state.
The state saves as a JSON file under document folder (e.g. `/UI_state/<user id>/ConversationListViewModel.json`).
When the collapse state of a folder changes, the state is refreshed and stored.
When `ConversationListViewModel` is created or the folder mode is togged, the state is read and set the collapse state if necessary. 
`ConversationListBottomBarController` 's selected button is set when `ConversationListViewModel`'s `restorationDelegate` is set.

### Todo

- [x] Test for state saving/restoring 
- [ ] move `wr_excludeFromBackup` method to `utility`

### CODE REVIEW

File: Wire-iOS/Sources/Helpers/FileDirectory/FileManager+Directory.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+private let zmLog = ZMSLog(tag: "FileManager")
+
+extension FileManager {
+
+    /// Create a directiory excluded from back up
+    ///
+    /// - Parameter pathComponent: folder to create
+    @discardableResult
+    func createBackupExcludedDirectoryIfNeeded(_ pathComponent: String) -> URL? {
```
Comment (typfel) [inline Wire-iOS/Sources/Helpers/FileDirectory/FileManager+Directory.swift]: We can't use the `createAndProtectDirectory` from `wire-ios-utilities`?

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift [hunk 2]
```diff
@@ -499,8 +546,102 @@ final class ConversationListViewModel: NSObject {
...
+        }
+
+        guard let state = state else { return }
+
+        folderEnabled = state.folderEnabled
+
+        restoreCollpasion()
+    }
+
+    private func restoreCollpasion() {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift]: typo?

File: Wire-iOS/Sources/Helpers/FileDirectory/FileManager+Directory.swift (hunk 1 above)
Comment (billypchan) [reply Wire-iOS/Sources/Helpers/FileDirectory/FileManager+Directory.swift]: I will clean up some duplicated methods form UI project to `wire-ios-utilities`

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift [hunk 3]
```diff
@@ -125,7 +125,7 @@ final class ConversationListViewModel: NSObject {
 
             /// restore collapse state
             if folderEnabled {
-                restoreCollpasion()
+                restoreCollpase()
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift]: typo: `restoreCollpase` >  `restoreCollapse`

File: Wire-iOS Tests/ConversationList/ConversationListViewModel/ConversationListViewModelTests.swift [hunk 4]
```diff
@@ -220,4 +223,80 @@ final class ConversationListViewModelTests: XCTestCase {
         XCTAssertEqual(sut.selectItem(at: indexPath), mockConversation)
     }
 
+    // MARK: - state
+    func testForCollapseState() {
```
Comment (typfel) [inline Wire-iOS Tests/ConversationList/ConversationListViewModel/ConversationListViewModelTests.swift]: It's hard to read tests which test multiple scenarios in one test. Can we try split up this test into multiple tests with descriptive names about what scenario they are testing.

Comment (billypchan) [reply Wire-iOS Tests/ConversationList/ConversationListViewModel/ConversationListViewModelTests.swift]: renamed to `testThatCollapseStateCanBeRestoredAfterFolderDisabled`

File: Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift [hunk 5]
```diff
@@ -499,8 +546,104 @@ final class ConversationListViewModel: NSObject {
...
+
+        do {
+            try jsonString.write(to: directoryURL.appendingPathComponent(ConversationListViewModel.persistentFilename), atomically: true, encoding: .utf8)
+        } catch {
+            log.error("error writing ConversationListViewModel to \(directoryURL): \(error)")
+        }
+    }
+
+    private func restoreState() {
+        if state == nil {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift]: Couldn't this be replace with a `guard state == nil else { return }`

Comment (billypchan) [reply Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift]: refactor `state` to a`lazy var`

**Pull request merged**